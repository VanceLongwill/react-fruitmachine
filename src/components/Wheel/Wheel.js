// @flow
import React from 'react'
import styled, { keyframes } from 'styled-components'

import {
  ANIMATION_DURATION,
  WHEEL_ANGLE,
  WHEEL_RADIUS,
  WHEELFACES_PER_ROW,
  COLORS_BY_INDEX,
} from '../../config/constants'
import WheelFace from '../WheelFace'

function getPosition(index) {
  return index * WHEEL_ANGLE + 360 * 2
}

const AnimatedWheel = styled.div`
  height: 120px;
  transform-style: preserve-3d;
  animation-iteration-count: 1;
  animation-timing-function: linear;
  animation-duration: ${ANIMATION_DURATION}s;
  animation-fill-mode: forwards;
  animation-name: ${p => p.animation};
  animation-delay: -${p => p.delay}ms;
`
type Props = {
  isLoading: boolean,
  delay: number,
  prevIndex: number,
  nextIndex: number,
}
export function Wheel(props: Props) {
  const { isLoading, delay, prevIndex, nextIndex } = props

  let toPos = nextIndex !== -1 ? getPosition(nextIndex) : 360
  if (prevIndex === nextIndex) {
    // make the wheel spin even if it will land at the same place twice
    toPos += 360
  }

  const animation = keyframes`
    from {
      transform: rotateX(0deg)
    }
    to {
      transform: rotateX(-${toPos}deg)
    }
  `

  const wheelFaces = []
  for (let i = 0; i < WHEELFACES_PER_ROW; i++) {
    wheelFaces.push(
      <WheelFace
        key={`${i}-${delay}`}
        rotateX={WHEEL_ANGLE * i}
        radius={WHEEL_RADIUS}
        color={COLORS_BY_INDEX[i].value}
      />
    )
  }

  return (
    <AnimatedWheel
      isLoading={isLoading}
      delay={delay}
      animation={animation}>
      {wheelFaces}
    </AnimatedWheel>
  )
}
