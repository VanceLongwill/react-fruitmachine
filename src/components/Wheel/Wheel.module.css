.wheel {
  margin: 0 auto;
  height: 110px;
  width: 600px;
  transform-style: preserve-3d;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
  animation-name: x-spin;
  animation-duration: 2s;
}

@keyframes x-spin {
  0%    { transform: rotateX(0deg); }
  50%   { transform: rotateX(180deg); }
  100%  { transform: rotateX(360deg); }
}

.wheel > :nth-child(odd) {
  background-color: #995C7F;
}

.wheel > :nth-child(even) {
  background-color: #835A99;
}

/*
.Wheel:first-child {
  z-index:-1;
  animation: image-change 10s ease infinite 1s;
}

@keyframes image-change {
 0%,50% {
    transform: translateY(0%);
  }
  70%,100% {
    transform: translateY(400%);
  }
}
*/
