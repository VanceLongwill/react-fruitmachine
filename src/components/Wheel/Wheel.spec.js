import React from 'react'
import { mount } from 'enzyme'

import { Wheel } from './Wheel'

describe('Wheel', () => {
  it('should match snapshot on initial render', () => {
    const props = {
      isLoading: false,
      delay: 300,
      prevIndex: 0,
      nextIndex: -1,
    }
    expect(mount(<Wheel {...props}/>)).toMatchSnapshot()
  })
  it('should match snapshot', () => {
    const props = {
      isLoading: false,
      delay: 300,
      prevIndex: 1,
      nextIndex: 7,
    }
    expect(mount(<Wheel {...props}/>)).toMatchSnapshot()
  })
})
