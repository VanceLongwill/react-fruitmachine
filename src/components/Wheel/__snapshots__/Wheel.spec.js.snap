// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wheel should match snapshot 1`] = `
<styled.div
  animation={
    Keyframes {
      "id": "sc-keyframes-fbSDaz",
      "inject": [Function],
      "name": "fbSDaz",
      "rules": Array [
        "@-webkit-keyframes fbSDaz{from{-webkit-transform:rotateX(0deg);-ms-transform:rotateX(0deg);transform:rotateX(0deg);}to{-webkit-transform:rotateX(-930deg);-ms-transform:rotateX(-930deg);transform:rotateX(-930deg);}}",
        "@keyframes fbSDaz{from{-webkit-transform:rotateX(0deg);-ms-transform:rotateX(0deg);transform:rotateX(0deg);}to{-webkit-transform:rotateX(-930deg);-ms-transform:rotateX(-930deg);transform:rotateX(-930deg);}}",
      ],
      "toString": [Function],
    }
  }
  delay={300}
  isLoading={false}
>
  <div
    className="sc-bwzfXH kFnvCg"
  >
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={0}
    >
      <div
        className="sc-bdVaJa eIXeZU"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={30}
    >
      <div
        className="sc-bdVaJa VTciH"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={60}
    >
      <div
        className="sc-bdVaJa eBVXud"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={90}
    >
      <div
        className="sc-bdVaJa dDjJrr"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={120}
    >
      <div
        className="sc-bdVaJa eanJqt"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={150}
    >
      <div
        className="sc-bdVaJa eonNfq"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={180}
    >
      <div
        className="sc-bdVaJa gPqlbN"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={210}
    >
      <div
        className="sc-bdVaJa esVdTs"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={240}
    >
      <div
        className="sc-bdVaJa dNkfLi"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={270}
    >
      <div
        className="sc-bdVaJa lfcSTa"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={300}
    >
      <div
        className="sc-bdVaJa bVdIgK"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={330}
    >
      <div
        className="sc-bdVaJa gWhcpB"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
  </div>
</styled.div>
`;

exports[`Wheel should match snapshot on initial render 1`] = `
<styled.div
  animation={
    Keyframes {
      "id": "sc-keyframes-iffaHa",
      "inject": [Function],
      "name": "iffaHa",
      "rules": Array [
        "@-webkit-keyframes iffaHa{from{-webkit-transform:rotateX(0deg);-ms-transform:rotateX(0deg);transform:rotateX(0deg);}to{-webkit-transform:rotateX(-360deg);-ms-transform:rotateX(-360deg);transform:rotateX(-360deg);}}",
        "@keyframes iffaHa{from{-webkit-transform:rotateX(0deg);-ms-transform:rotateX(0deg);transform:rotateX(0deg);}to{-webkit-transform:rotateX(-360deg);-ms-transform:rotateX(-360deg);transform:rotateX(-360deg);}}",
      ],
      "toString": [Function],
    }
  }
  delay={300}
  isLoading={false}
>
  <div
    className="sc-bwzfXH UCihG"
  >
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={0}
    >
      <div
        className="sc-bdVaJa eIXeZU"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={30}
    >
      <div
        className="sc-bdVaJa VTciH"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={60}
    >
      <div
        className="sc-bdVaJa eBVXud"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={90}
    >
      <div
        className="sc-bdVaJa dDjJrr"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={120}
    >
      <div
        className="sc-bdVaJa eanJqt"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={150}
    >
      <div
        className="sc-bdVaJa eonNfq"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={180}
    >
      <div
        className="sc-bdVaJa gPqlbN"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={210}
    >
      <div
        className="sc-bdVaJa esVdTs"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(198, 6, 0)"
      radius={200}
      rotateX={240}
    >
      <div
        className="sc-bdVaJa dNkfLi"
        color="rgb(198, 6, 0)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(105, 150, 15)"
      radius={200}
      rotateX={270}
    >
      <div
        className="sc-bdVaJa lfcSTa"
        color="rgb(105, 150, 15)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(79, 190, 242)"
      radius={200}
      rotateX={300}
    >
      <div
        className="sc-bdVaJa bVdIgK"
        color="rgb(79, 190, 242)"
        radius={200}
      />
    </styled.div>
    <styled.div
      color="rgb(249, 244, 87)"
      radius={200}
      rotateX={330}
    >
      <div
        className="sc-bdVaJa gWhcpB"
        color="rgb(249, 244, 87)"
        radius={200}
      />
    </styled.div>
  </div>
</styled.div>
`;
